### PR TITLE
[ISSUE #10976] Reduce per-message allocation on the proxy gRPC path

### LINKED ISSUE #10976: [Enhancement] Avoid per-message MessageDigest lookup and length-only getBytes allocations on the proxy gRPC path
### Before Creating the Enhancement Request

- [x] I have confirmed that this should be classified as an enhancement rather than a bug/feature.

### Summary

Remove two per-message allocation sources on the proxy gRPC path: the `MessageDigest.getInstance("MD5")` lookup for every delivered message, and the `getBytes(UTF_8)` calls used only for length validation on every received message.

### Motivation

1. `GrpcConverter#buildSystemProperties` computes the body digest for every message delivered to a gRPC consumer, and `BinaryUtil.calculateMd5` performs a `MessageDigest.getInstance("MD5")` provider lookup plus a fresh digest instance on every call.
2. `SendMessageActivity#buildMessageProperty` (and `validateMessageGroup`) call `str.getBytes(StandardCharsets.UTF_8)` on every user-property key/value, the tag, each message key, and the message group — only to read `.length` for size validation; the byte arrays are discarded immediately. That is 2N+ transient arrays per received message (N = user property count).

### Solution

- `BinaryUtil`: keep one `MessageDigest` per thread in a `ThreadLocal` (`MessageDigest` is not thread safe) and `reset()` before each use.
- `SendMessageActivity`: add a `utf8Length(String)` helper that computes the UTF-8 encoded length without materializing the array, matching `String.getBytes(UTF_8).length` exactly, including the single-byte replacement for unpaired surrogates; replace the five call sites.

### Verification

- `BinaryUtilTest` (new): digest matches a fresh `MessageDigest` and stays stable across interleaved calls on the reused per-thread instance. `SendMessageActivityTest#testUtf8Length`: sample-by-sample equality with `getBytes(UTF_8).length` covering ASCII, CJK, supplementary (emoji), and unpaired surrogates; full class 12/12.
- Dedicated gRPC A/B on a 4-node cluster: proxy (cluster mode) plus a loopback load tool built on `rocketmq-client-java` 5.0.7 (producer 8 threads with multi-byte user properties exercising `utf8Length`, SimpleConsumer 4 threads exercising the digest path), swapping the proxy's `rocketmq-common`/`rocketmq-proxy` jars per arm; 3 interleaved trials plus 1 reversed-order control: ~9.4k send TPS / ~5.1k consume TPS, zero failures in all 8 arms; proxy young GC showed a pure positional artifact (first arm of each pair always 9, second always 10, independent of the jar — confirmed by the reversed-order control), i.e. parity after correction. No regression; the allocation saving itself is below GC-count resolution, so this is a cleanup-level optimization on the proxy hot path.

### PR DESCRIPTION
**Pull request by wang-jiahua**
### Which Issue(s) This PR Fixes

Fixes #10976

### Brief Description

Two per-message allocation sources removed on the proxy gRPC path:

1. `BinaryUtil.calculateMd5` performed a `MessageDigest.getInstance("MD5")` provider lookup and created a fresh digest on every call; `GrpcConverter#buildSystemProperties` hits this for every message delivered to a gRPC consumer. The digest is now kept per thread in a `ThreadLocal` (`MessageDigest` is not thread safe) and `reset()` before each use.
2. `SendMessageActivity#buildMessageProperty` (and `validateMessageGroup`) called `getBytes(StandardCharsets.UTF_8)` on every user-property key/value, the tag, each message key, and the message group, only to read `.length` for size validation. A new `utf8Length(String)` helper computes the encoded length without materializing the array, matching `String.getBytes(UTF_8).length` exactly (including the single-byte replacement for unpaired surrogates); the five call sites now use it.

### How Did You Test This Change?

- `BinaryUtilTest` (new): result matches a fresh `MessageDigest`, and stays stable across interleaved calls on the reused per-thread instance.
- `SendMessageActivityTest#testUtf8Length`: sample-by-sample equality with `getBytes(UTF_8).length` covering ASCII, CJK, supplementary (emoji), and unpaired surrogates; full class 12/12.
- Dedicated gRPC A/B on a 4-node cluster: proxy in cluster mode plus a loopback load tool built on `rocketmq-client-java` 5.0.7 (producer 8 threads with multi-byte user properties exercising `utf8Length`; SimpleConsumer 4 threads exercising the digest path), swapping the proxy's `rocketmq-common`/`rocketmq-proxy` jars per arm, 3 interleaved trials plus 1 reversed-order control: ~9.4k send TPS / ~5.1k consume TPS with zero failures in all 8 arms; proxy young GC counts showed a pure positional artifact (first arm of each pair always 9, second always 10, independent of the jar — confirmed by the reversed-order control), i.e. parity after correction. No regression; the allocation saving is below GC-count resolution, so this is a cleanup-level optimization on the proxy hot path.

### CODE REVIEW

**Review by RockteMQ-AI — APPROVED** [review]
﻿### Summary
This PR eliminates two per-message allocation sources on the proxy gRPC path: (1) `MessageDigest.getInstance("MD5")` provider lookup on every `calculateMd5` call, replaced with a `ThreadLocal<MessageDigest>` that is `reset()` before each use; (2) `str.getBytes(UTF_8).length` for measuring UTF-8 encoded size, replaced with a branch-based `utf8Length()` that computes the byte count without allocating a temporary array.

Both changes are correct and well-motivated. The `ThreadLocal` pattern for `MessageDigest` is standard practice for non-thread-safe reusable objects, and `reset()` is properly called before each use. The `utf8Length()` implementation correctly handles all cases: ASCII, BMP, supplementary characters (surrogate pairs), and unpaired surrogates (matching Java's `getBytes(UTF_8)` replacement behavior). Tests cover all edge cases including interleaved payloads to verify the ThreadLocal does not corrupt results.

LGTM — clean, focused performance optimization with solid test coverage.

---
*Automated review by github-manager-bot*

File: common/src/main/java/org/apache/rocketmq/common/utils/BinaryUtil.java [hunk 1]
```diff
@@ -23,13 +23,21 @@
...
+    /**
+     * MessageDigest is not thread safe, so keep one instance per thread instead of
+     * looking it up from the provider on every call.
+     */
+    private static final ThreadLocal<MessageDigest> MD5_DIGEST = ThreadLocal.withInitial(() -> {
         try {
-            messageDigest = MessageDigest.getInstance("MD5");
+            return MessageDigest.getInstance("MD5");
         } catch (NoSuchAlgorithmException e) {
             throw new RuntimeException("MD5 algorithm not found.");
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply common/src/main/java/org/apache/rocketmq/common/utils/BinaryUtil.java]: Fixed — the RuntimeException now carries the original NoSuchAlgorithmException as its cause (commit a7df129).

File: proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/producer/SendMessageActivity.java [hunk 2]
```diff
@@ -311,6 +310,35 @@ protected Map<String, String> buildMessageProperty(ProxyContext context, apache.
         return messageWithHeader.getProperties();
     }
 
+    /**
+     * Length of the UTF-8 encoding of {@code str} without materializing the byte array.
+     * Matches {@code str.getBytes(StandardCharsets.UTF_8).length}, including the
+     * single-byte replacement for unpaired surrogates.
+     */
+    protected static int utf8Length(String str) {
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/producer/SendMessageActivity.java]: Narrowed to package-private — the test lives in the same package, so no need to expose it to subclasses (commit a7df129).

**Review by RockteMQ-AI — APPROVED** [review]
### Summary
Reduces per-message allocation on the proxy gRPC path by (1) caching `MessageDigest` in a `ThreadLocal` in `BinaryUtil.calculateMd5` and (2) replacing `String.getBytes(UTF_8).length` with a zero-allocation `utf8Length()` in `SendMessageActivity`.

Changes look correct:
- `ThreadLocal<MessageDigest>` with `reset()` before each use is the right pattern for non-thread-safe digest instances
- Test interleaves calls to verify reset stability — good coverage
- `utf8Length()` avoids byte array allocation for size validation on the hot path

Minor note: the `.gitignore` additions (`.qoder/`, `profiling-tools/`, etc.) appear to be local dev artifacts — consider whether these belong in a project-wide `.gitignore`.

LGTM.

---
*Automated review by github-manager-bot*
